From 39581a57709888c307117ff89e06be67a245fc08 Mon Sep 17 00:00:00 2001
From: Jordan Bayles <jophba@chromium.org>
Date: Tue, 23 Apr 2024 21:38:39 +0000
Subject: [PATCH] [Cast Message Util Fuzzer] fix reproducible crash

This patch fixes a JSON parsing crash in the cast_message_util_fuzzer.

Bug: 327486176
Change-Id: I0abb42b1f56db155910b9f21180757bd17798b97
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5463569
Commit-Queue: Jordan Bayles <jophba@chromium.org>
Reviewed-by: Mark Foltz <mfoltz@chromium.org>
Reviewed-by: Takumi Fujimoto <takumif@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1291552}
---

--- a/components/media_router/common/providers/cast/channel/cast_message_util_fuzzer.cc
+++ b/components/media_router/common/providers/cast/channel/cast_message_util_fuzzer.cc
@@ -19,16 +19,25 @@
 
 namespace {
 
+base::Value MakeValue(const JunkValue::Field& field) {
+  if (field.has_int_value()) {
+    return base::Value(field.int_value());
+  }
+  if (field.has_string_value()) {
+    return base::Value(field.string_value());
+  }
+  if (field.has_float_value()) {
+    return std::isfinite(field.float_value()) ? base::Value(field.float_value())
+                                              : base::Value();
+  }
+  CHECK(field.has_bool_value());
+  return base::Value(field.bool_value());
+}
+
 base::Value::Dict MakeDict(const JunkValue& junk) {
   base::Value::Dict result;
-  for (int i = 0; i < junk.field_size(); i++) {
-    const auto& field = junk.field(i);
-    base::Value field_value =
-        field.has_int_value()      ? base::Value(field.int_value())
-        : field.has_string_value() ? base::Value(field.string_value())
-        : field.has_float_value()  ? base::Value(field.float_value())
-                                   : base::Value(field.bool_value());
-    result.Set(field.name(), std::move(field_value));
+  for (const auto& field : junk.field()) {
+    result.Set(field.name(), MakeValue(field));
   }
   return result;
 }
